### PR TITLE
Use field.Error(s) in scheduler plugin args validation

### PR DESCRIPTION
**Pull request by lingsamuel**
Signed-off-by: Ling Samuel <lingsamuelgrace@gmail.com>

**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Provides all errors detected instead of first one.

**Which issue(s) this PR fixes**:

Could be part of #96167

**Special notes for your reviewer**:

1. We could return `allErrs.ToAggregate()` directly because it checks `len(allErrs)==0`
2. Exported function signature must use `error` interface as return type. Because cast a pointer to interface could make `== nil` return false.

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by alculquicondor** [comment]
/assign @damemi 
for a first pass. No rush though.

File: pkg/scheduler/apis/config/validation/validation_pluginargs.go
Comment (damemi) [inline pkg/scheduler/apis/config/validation/validation_pluginargs.go:96]: This might be better renamed to something that indicates it's specifically for validating present/absent label conflicts, but I wouldn't block on that

File: pkg/scheduler/apis/config/validation/validation_pluginargs.go
Comment (damemi) [inline pkg/scheduler/apis/config/validation/validation_pluginargs.go:213]: any reason to break this into a new function instead of keeping all the logic here? You could still append the errors based on the index right?

File: pkg/scheduler/apis/config/validation/validation_pluginargs.go
Comment (damemi) [inline pkg/scheduler/apis/config/validation/validation_pluginargs.go:271]: same question as above, does this need to be a new function?

**Review by damemi — COMMENTED** [review]
looks pretty good, but just a couple questions on some of the new functions

File: pkg/scheduler/apis/config/validation/validation_pluginargs.go
Comment (lingsamuel) [reply pkg/scheduler/apis/config/validation/validation_pluginargs.go:96]: This function name is not created by me. But I think such function name is enough because if we want to present it's functionality (validating present/absent label conflicts), it may be too long. Comment above the function already says it.

File: pkg/scheduler/apis/config/validation/validation_pluginargs.go
Comment (lingsamuel) [reply pkg/scheduler/apis/config/validation/validation_pluginargs.go:213]: No special reason, it's just my habit. I think it would be more clearer to create a function to process a single value instead of embedding it in a list process function. But if you think the previous version could be better, I can revert it.

File: pkg/scheduler/apis/config/validation/validation_pluginargs.go
Comment (lingsamuel) [inline pkg/scheduler/apis/config/validation/validation_pluginargs.go:207]: I have some concerns here. If there are many unsorted elements, then "utilization values must be sorted." will be repeated many times.

Comment (damemi) [reply pkg/scheduler/apis/config/validation/validation_pluginargs.go:214]: If you want, you could just identify the first one and break from the loop like before. But if you want to return all the errors I don't see a much better way to collect them other than maybe building a list of all the invalid fields and combining them into 1 error.

**Review by damemi — COMMENTED** [review]
/lgtm
I'll leave it up to you if you want to try to refactor that check for unsorted values. It's also something that could be revisited if it ends up being very spammy

**Comment by lingsamuel** [comment]
My first version like this:
```diff
+	var unsortedErrs field.ErrorList
 	for i := 1; i < len(shape); i++ {
 		if shape[i-1].Utilization >= shape[i].Utilization {
-			msg := fmt.Sprintf("utilization values must be sorted. Should < %s==%d", p.Index(i).Child("utilization").String(), shape[i].Utilization)
-			allErrs = append(allErrs, field.Invalid(p.Index(i-1).Child("utilization"), shape[i-1].Utilization, msg))
+			msg := fmt.Sprintf("should < %s==%d", p.Index(i).Child("utilization").String(), shape[i].Utilization)
+			unsortedErrs = append(unsortedErrs, field.Invalid(p.Index(i-1).Child("utilization"), shape[i-1].Utilization, msg))
 		}
 	}
+	capFirstChar := func (s string) string {
+		for index, value := range s {
+			return string(unicode.ToUpper(value)) + s[index+1:]
+		}
+		return ""
+	}
+
+	if len(unsortedErrs) == 1 {
+		unsortedErrs[0].Detail = "utilization values must be sorted. " + capFirstChar(unsortedErrs[0].Detail)
+		allErrs = append(allErrs, unsortedErrs...)
+	} else if len(unsortedErrs) > 1 {
+		var utilizations []int32
+		for _, p := range shape {
+			utilizations = append(utilizations, p.Utilization)
+		}
+		allErrs = append(allErrs, field.Invalid(p.Key("*").Child("utilization"), fmt.Sprintf("%+v", utilizations), "utilization values must be sorted"))
+		allErrs = append(allErrs, unsortedErrs...)
+	}
 
 	for i, point := range shape {
 		utilizationPath := p.Index(i).Child("utilization")
```

It produces msg like:
` [shape[*].utilization: Invalid value: "[30 20 10]": utilization values must be sorted, shape[0].utilization: Invalid value: 30: should < shape[1].utilization==20, shape[1].utilization: Invalid value: 20: should < shape[2].utilization==10] `.

This patch prepends the message "utilization values must be sorted" before all detailed errors.
But two questions here:
1. `shape[*].utilization` maybe not a standard format
2. Too complex. The error case is rare. I don't think it's worth it.

So IMO we could ignore this problem at this time. In case we need it, we could apply the patch above.

File: pkg/scheduler/apis/config/validation/validation_pluginargs.go
Comment (alculquicondor) [inline pkg/scheduler/apis/config/validation/validation_pluginargs.go:31]: Eventually, we will make this function accept a path #88093

In preparation for that do the following:

```golang
var path field.Path
if err := validateMinCandidateNodesPercentage(args.MinCandidateNodesPercentage, path.Child("minCandidateNodesPercentage")) {
```

Do the same for all the functions

File: pkg/scheduler/apis/config/validation/validation_pluginargs.go
Comment (alculquicondor) [inline pkg/scheduler/apis/config/validation/validation_pluginargs.go:41]: Add 2 errors, one for each field.

File: pkg/scheduler/apis/config/validation/validation_pluginargs.go
Comment (alculquicondor) [inline pkg/scheduler/apis/config/validation/validation_pluginargs.go:96]:
```suggestion
func validateNoConflict(presentLabels, absentLabels []string, presentPath, absentPath *field.Path) field.ErrorList {
```

File: pkg/scheduler/apis/config/validation/validation_pluginargs.go
Comment (alculquicondor) [inline pkg/scheduler/apis/config/validation/validation_pluginargs.go:197]: nit: use `path` as variable name. The function is somewhat long.

File: pkg/scheduler/apis/config/validation/validation_pluginargs.go
Comment (alculquicondor) [reply pkg/scheduler/apis/config/validation/validation_pluginargs.go:214]: Error on the first one is fine.

No need to mention the values

```golang
field.Invalid(p.Index(i).Child("utilization"), shape[i].Utilization, "utilization values must be sorted in increasing order")
```

File: pkg/scheduler/apis/config/validation/validation_pluginargs.go
Comment (alculquicondor) [inline pkg/scheduler/apis/config/validation/validation_pluginargs.go:230]: For consistency with other messages, make the detail:

`not in the range [x, y]`

File: pkg/scheduler/apis/config/validation/validation_pluginargs.go
Comment (alculquicondor) [inline pkg/scheduler/apis/config/validation/validation_pluginargs.go:267]: same, use a range in the detail

File: pkg/scheduler/apis/config/validation/validation_pluginargs_test.go
Comment (alculquicondor) [inline pkg/scheduler/apis/config/validation/validation_pluginargs_test.go:361]: let's get rid of string comparisons. See #96522

File: pkg/scheduler/apis/config/validation/validation_pluginargs.go
Comment (lingsamuel) [reply pkg/scheduler/apis/config/validation/validation_pluginargs.go:41]: Like this?
```go
allErrs = append(allErrs, field.Invalid(field.NewPath("minCandidateNodesPercentage"), args.MinCandidateNodesPercentage, "cannot be zero at the same time as minCandidateNodesAbsolute"))
allErrs = append(allErrs, field.Invalid(field.NewPath("minCandidateNodesAbsolute"), args.MinCandidateNodesAbsolute, "cannot be zero at the same time as minCandidateNodesPercentage"))
```

Comment (alculquicondor) [reply pkg/scheduler/apis/config/validation/validation_pluginargs.go:41]: Yes, but you can do that in a single `append`

Comment (lingsamuel) [reply pkg/scheduler/apis/config/validation/validation_pluginargs.go:41]: Perhaps it should be repeated twice semantically. But I think this is a bit redundant.

File: pkg/scheduler/apis/config/validation/validation_pluginargs_test.go
Comment (lingsamuel) [inline pkg/scheduler/apis/config/validation/validation_pluginargs_test.go:929]: I am no using this func in `TestValidateNodeAffinityArgs` because it was changed at #96522

Comment (alculquicondor) [reply pkg/scheduler/apis/config/validation/validation_pluginargs_test.go:929]: Remove this function. Use `Diff` in the tests directly.

**Review by damemi — COMMENTED** [review]
/ok-to-test
(didn't realize this still needed it, sorry)

**Comment by damemi** [comment]
@lixiaobing1 it looks like these bazel failures are coming from `pkg/scheduler/framework/plugins/noderesources/least_allocated_test.go`? Maybe you need to update the `noderesources` tests similarly

**Comment by damemi** [comment]
> @lixiaobing1 it looks like these bazel failures are coming from `pkg/scheduler/framework/plugins/noderesources/least_allocated_test.go`? Maybe you need to update the `noderesources` tests similarly

sorry, tagged the wrong person... @lingsamuel :)

**Comment by lingsamuel** [comment]
done

File: pkg/scheduler/framework/plugins/noderesources/least_allocated_test.go
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/noderesources/least_allocated_test.go:107]: this should be `wantErr error`

File: pkg/scheduler/framework/plugins/noderesources/most_allocated_test.go
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/noderesources/most_allocated_test.go:122]: same, `wantErr error`

File: pkg/scheduler/apis/config/validation/validation_pluginargs_test.go
Comment (alculquicondor) [inline pkg/scheduler/apis/config/validation/validation_pluginargs_test.go:37]: `wantErr error`

File: pkg/scheduler/apis/config/validation/validation_pluginargs_test.go
Comment (alculquicondor) [inline pkg/scheduler/apis/config/validation/validation_pluginargs_test.go:166]: `wantErr error`

**Comment by lingsamuel** [comment]
It seems that I can no longer make this e2e test pass.
But I don't know what affected the e2e test [between these two commits](https://github.com/lingsamuel/kubernetes/compare/d8ff935222a2602d00eaf389445a0b5517378517..77f0f442a187c81dbf15e7327219608074474874) :(.

**Review by alculquicondor — COMMENTED** [review]
/approve

leaving lgtm to @damemi

**Review by damemi — APPROVED** [review]
/lgtm

**Comment by damemi** [comment]
The test failures are an unrelated flake that's being worked on (see https://github.com/kubernetes/kubernetes/issues/97002)

**Comment by lingsamuel** [comment]
Do I need to restart it again and again until it succeeds?

**Comment by lingsamuel** [comment]
/hold
due to the test can't pass

**Comment by damemi** [comment]
@lingsamuel the test failures aren't related to your changes, it's a flake that's being worked on by some other people. if you want to hold to stop the auto-retest spam that's fine, just making sure you know :)

**Comment by alculquicondor** [comment]
Actually, thanks for holding. We don't want to stress the test queue.

**Comment by lingsamuel** [comment]
:thinking: why the test reran

**Pull request merged**
